### ISSUE TITLE
Building mmark for Raspberry Pi under Go 1.5; inconsistent results

### ISSUE THREAD

**Issue by vielmetti** [miekg/mmark#17]
I was able to cross-compile mmark for my Raspberry Pi 2 and get a working binary, using Go 1.5 on my Mac.

I followed the examples here:

http://labs.yulrizka.com/en/2015/08/testing-go-1-dot-5-cross-compilation.html

with this command line

```
env GOOS=linux GOARCH=arm GOARM=6 go build
```

that builds an image that will run on the Pi and the Pi 2, and

```
env GOOS=linux GOARCH=arm GOARM=7 go build
```

that builds an image just for the Pi 2.  You can then move the binary you just built over to the PI and run it.

However.....

When I build RFC 3514 using this command line

```
time ./mmark.arm7 -xml2 -page ../tmp/rfc3514.md > ~/tmp/3514.xml.7
time ./mmark.arm6 -xml2 -page ../tmp/rfc3514.md > ~/tmp/3514.xml.6
```

I do not get identical results, and when I further compare with the Darwin binary there are other differences. They are minor -- the section

```
  <references title="Normative References">
```

puts the references in a slightly different order. I think that's probably because of the "range" operator at https://github.com/miekg/mmark/blob/master/xml2rfcv2.go#L547 

```
                        for _, c := range citations {
```

doesn't guarantee repeated consistent results. (This refers directly to "integration test" considerations previously mentioned in #8).

> The iteration order over maps is not specified and is not guaranteed to be the same from one iteration to the next. If map entries that have not yet been reached are removed during iteration, the corresponding iteration values will not be produced. If map entries are created during iteration, that entry may be produced during the iteration or may be skipped. The choice may vary for each entry created and from one iteration to the next. If the map is nil, the number of iterations is 0.

So to summarize!

1. Works under Go 1.5 !
2. Works when cross-compiled under ARM 6 and 7 on Raspberry Pi 2 !
3. References sometimes appear in a different order in the XML2 output, which should be fixable with a stable sort !

**Comment by miekg** [comment]
cool!

Added sorting of the refs in 1c7e5aa23f8019b345c78ade59a891531b67b162

**Comment by miekg** [comment]
Think this can be closed now? (i.e. fix is in)

**Comment by vielmetti** [comment]
Yup, OK to close.

On Wed, Sep 9, 2015 at 3:55 PM, Miek Gieben <notifications@github.com>
wrote:

> Think this can be closed now? (i.e. fix is in)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/miekg/mmark/issues/17#issuecomment-139027903>.
>



-- 
Edward Vielmetti +1 734 330 2465
edward.vielmetti@gmail.com
